indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java

```diff
@@ -317,6 +322,157 @@ public void handle()
+    int taskCountMin = Integer.parseInt(String.valueOf(dynamicAllocationTasksProperties.getOrDefault("taskCountMin", 1)));
+    int scaleInStep = Integer.parseInt(String.valueOf(dynamicAllocationTasksProperties.getOrDefault("scaleInStep", 1)));
+    int scaleOutStep = Integer.parseInt(String.valueOf(dynamicAllocationTasksProperties.getOrDefault("scaleOutStep", 2)));
+    int beyond = 0;
+    int within = 0;
+    int metricsCount = lags.size();
+    for (Long lag : lags) {
+      if (lag >= scaleOutThreshold) {
+        beyond++;
+      }
+      if (lag <= scaleInThreshold) {
+        within++;
+      }
+    }
+    double beyondProportion = beyond * 1.0 / metricsCount;
+    double withinProportion = within * 1.0 / metricsCount;
+    log.info("triggerSaleOutThresholdFrequency is [ " + triggerSaleOutThresholdFrequency + " ] and triggerSaleInThresholdFrequency is [ " + triggerSaleInThresholdFrequency + " ]");
+    log.info("beyondProportion is [ " + beyondProportion + " ] and withinProportion is [ " + withinProportion + " ]");
+
+    int currentActiveTaskCount;
+    int desireActiveTaskCount;
+    Collection<TaskGroup> activeTaskGroups = activelyReadingTaskGroups.values();
+    currentActiveTaskCount = activeTaskGroups.size();
+
+    if (beyondProportion >= triggerSaleOutThresholdFrequency) {
+      // Do Scale out
+      int taskCount = currentActiveTaskCount + scaleOutStep;
+      if (currentActiveTaskCount == taskCountMax) {
+        log.info("CurrentActiveTaskCount reach task count Max limit, skip to scale out tasks");
+        return false;
+      } else {
+        desireActiveTaskCount = Math.min(taskCount, taskCountMax);
+      }
+      log.info("Start to scale out tasks , current active task number [ " + currentActiveTaskCount + " ] and desire task number is [ " + desireActiveTaskCount + " ] ");
+      gracefulShutdownInternal();
+      // clear everything
+      clearAllocationInfos();
+      log.info("Set Task Count : " + desireActiveTaskCount);
+      setTaskCount(desireActiveTaskCount);
+      return true;
+    }
+
+    if (withinProportion >= triggerSaleInThresholdFrequency) {
+      // Do Scale in
+      int taskCount = currentActiveTaskCount - scaleInStep;
+      if (currentActiveTaskCount == taskCountMin) {
+        log.info("CurrentActiveTaskCount reach task count Min limit, skip to scale in tasks");
+        return false;
+      } else {
+        desireActiveTaskCount = Math.max(taskCount, taskCountMin);
+      }
+      log.info("Start to scale in tasks , current active task number [ " + currentActiveTaskCount + " ] and desire task number is [ " + desireActiveTaskCount + " ] ");
+      gracefulShutdownInternal();
+      // clear everything
+      clearAllocationInfos();
+      log.info("Set Task Count : " + desireActiveTaskCount);
+      setTaskCount(desireActiveTaskCount);
+      return true;
+    }
+    return false;
+  }
+
+  private void setTaskCount(int desireActiveTaskCount)
```

Comment:
1. renamed as `changeTaskCountInIOConfig`
2. the method `gracefulShutdownInternal` will not suspend the supervisor. It will make the ingest tasks stop reading and start to publish data. So that whether the func fails or not, the supervisor is always active. What's more, if this func failed, current scale action will be canceled. And will try another scale action in `dynamicCheckPeriod`.